/*
 * Copyright 2018 source{d}. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
package pb;

import "github.com/gogo/protobuf/gogoproto/gogo.proto";
import "gopkg.in/bblfsh/sdk.v1/uast/generated.proto";
import "lookout/sdk/event.proto";

option (gogoproto.goproto_getters_all) = false;

// Data services exposes VCS repositories.
service Data {
    rpc GetChanges (ChangesRequest) returns (stream Change);
    rpc GetFiles (FilesRequest) returns (stream File);
}

message File {
    option (gogoproto.sizer) = false;
    option (gogoproto.protosizer) = true;

    // File path.
    string path = 1;
    // POSIX-style file mode.
    uint32 mode = 2;
    // Hash of the file contents.
    string hash = 3;
    // Raw content of the file.
    bytes content = 4;
    // UAST.
    gopkg.in.bblfsh.sdk.v1.uast.Node uast = 5 [(gogoproto.customname) = "UAST"];
    // Programming/data/markup language of the file.
    string language = 6;
}

message Change {
    option (gogoproto.sizer) = false;
    option (gogoproto.protosizer) = true;
    
    File base = 1;
    File head = 2;
}

message ChangesRequest {
    ReferencePointer base = 1;
    ReferencePointer head = 2;
    string include_pattern = 3;
    string exclude_pattern = 4;
    bool exclude_vendored = 5;
    bool want_contents = 6;
    bool want_uast = 7 [(gogoproto.customname) = "WantUAST"];
    bool want_language = 8;
    repeated string include_languages = 9;
}

message FilesRequest {
    ReferencePointer revision = 1;
    string include_pattern = 2;
    string exclude_pattern = 3;
    bool exclude_vendored = 4;
    bool want_contents = 5;
    bool want_uast = 6 [(gogoproto.customname) = "WantUAST"];
    // WantLanguage set to true if UAST was requested
    bool want_language = 7;
    repeated string include_languages = 8;
}
